deep_qa/training/text_trainer.py

```diff
@@ -173,7 +176,10 @@ def load_dataset_from_files(self, files: List[str]):
         has background information could call this method, then do additional processing on the
         rest of the list, for instance).
         """
-        return TextDataset.read_from_file(files[0], self._instance_type())
+        dataset_type_key = self.dataset_params.pop_choice("type", list(concrete_datasets.keys()),
+                                                          default_to_first_choice=True)
+        dataset_type = concrete_datasets[dataset_type_key]
+        return dataset_type.read_from_file(files[0], self._instance_type())
```

Comment:
This is why I like immutable state, instead of mutable state...  But the alternatives to having mutable state, while still doing easy parameter validation, aren't very good either.  Maybe the best option is to have `Params` keep track of which items have been read, and instead of what we currently do for `assert_empty`, change it to just check that all keys in the dictionary have actually been read.  A bit of a pain, but probably better than having these subtle bugs due to this mutable state.  That's for another PR, though, and I'm not sure how high priority it is...